### PR TITLE
feat: enable soft delete for resources

### PR DESCRIPTION
**Pull request by haritianaadriano**

### CODE REVIEW

File: src/main/resources/db/migration/V42_10__Update_resources_data.sql
Comment (Mahefaa) [inline src/main/resources/db/migration/V42_10__Update_resources_data.sql:1]: resources is too vague as a name

File: src/main/java/school/hei/haapi/repository/PaymentRepository.java [hunk 1]
```diff
@@ -13,7 +13,8 @@ public interface PaymentRepository extends JpaRepository<Payment, String> {
   @Query(
       value =
           "select p from Payment p join Fee f on f.id = p.fee.id"
-              + " where f.student.id = :student_id and f.id = :fee_id")
+              + " where f.student.id = :student_id and f.id = :fee_id"
+              + " and p.isDeleted = false")
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/repository/PaymentRepository.java]: test without adding isDeleted check in the sql request.

Normally the @Where decorator already checks this field's value

File: src/main/java/school/hei/haapi/repository/dao/CourseDao.java [hunk 2]
```diff
@@ -68,6 +68,8 @@ public List<Course> findByCriteria(
               builder.lower(teacher.get("firstName")), "%" + teacherFirstName.toLowerCase() + "%"));
     }
 
+    predicates.add(builder.isFalse(root.get("isDeleted")));
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/repository/dao/CourseDao.java]: not needed if the @Where decorator does it but test it anyway

File: src/main/java/school/hei/haapi/repository/dao/UserManagerDao.java [hunk 3]
```diff
@@ -50,6 +50,8 @@ public List<User> findByCriteria(
 
     Predicate hasUserRole = builder.equal(root.get("role"), role);
 
+    Predicate isNotDeleted = builder.isFalse(root.get("isDeleted"));
+
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/repository/dao/UserManagerDao.java]: same for this

File: src/main/resources/db/migration/V42_10__Update_resources_data.sql [hunk 4]
```diff
@@ -0,0 +1,14 @@
+-- alter table user
+alter table "user" add column if not exists is_deleted boolean default false;
+
+-- alter table payment
+alter table "payment" add column if not exists is_deleted boolean default false;
```
Comment (Mahefaa) [inline src/main/resources/db/migration/V42_10__Update_resources_data.sql]: delete payment ? ... seems a bit dubious but ok

File: src/main/java/school/hei/haapi/model/Course.java [hunk 5]
```diff
@@ -34,6 +38,8 @@ public class Course implements Serializable {
 
   private Integer totalHours;
 
+  private Boolean isDeleted;
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/model/Course.java]: why "Boolean" instead of boolean ? your fields  default to false !

**Review by Mahefaa — CHANGES_REQUESTED** [review]

File: doc/api.yml [hunk 6]
```diff
@@ -233,6 +233,34 @@ paths:
...
+      operationId: deleteStudentById
+      parameters:
+        - name: id
+          in: path
+          required: true
+          schema:
+            type: string
+      responses:
+        '200':
+          description: student deleted.
```
Comment (Mahefaa) [inline doc/api.yml]: deleted student

File: doc/api.yml [hunk 7]
```diff
@@ -233,6 +233,34 @@ paths:
           $ref: '#/components/responses/429'
         '500':
           $ref: '#/components/responses/500'
+    delete:
+      tags:
+        - Users
+      summary: Delete student by identifier.
```
Comment (Mahefaa) [inline doc/api.yml:239]: describe it as a soft delete

File: doc/api.yml [hunk 8]
```diff
@@ -1649,6 +1767,45 @@ paths:
         '500':
           $ref: '#/components/responses/500'
 
+  /students/{student_id}/fees/{fee_id}/payments/{payment_id}:
+    delete:
+      tags:
+        - Paying
+      summary: Delete student payment by identifier.
```
Comment (Mahefaa) [inline doc/api.yml]: delete student payment on a fee

File: doc/api.yml [hunk 9]
```diff
@@ -1649,6 +1767,45 @@ paths:
...
+      operationId: deleteStudentPaymentById
```
Comment (Mahefaa) [inline doc/api.yml]: deleteStudentFeePaymentById

File: src/main/java/school/hei/haapi/model/Course.java [hunk 10]
```diff
@@ -11,6 +11,8 @@
 import javax.persistence.OneToMany;
 import javax.persistence.Table;
 import lombok.*;
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/model/Course.java]: stop using import *

File: src/main/java/school/hei/haapi/model/Fee.java [hunk 11]
```diff
@@ -72,6 +76,8 @@ public class Fee implements Serializable {
   @OneToMany(mappedBy = "fee")
   private List<Payment> payments;
 
+  private boolean isDeleted = false;
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/model/Fee.java]: why ?

File: src/main/java/school/hei/haapi/service/CourseService.java [hunk 12]
```diff
@@ -50,6 +50,12 @@ public List<Course> getCourses(
         pageable);
   }
 
+  public Course deleteById(String courseId) {
+    Course course = getById(courseId);
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/CourseService.java]: why do you need to get it before deleting it ??

**Review by Mahefaa — CHANGES_REQUESTED** [review]

File: src/test/java/school/hei/haapi/integration/FeeIT.java [hunk 13]
```diff
@@ -59,6 +59,19 @@ void setUp() {
...
+  @DirtiesContext
+  void manager_delete_ok() throws ApiException {
+    ApiClient manager1Client = anApiClient(MANAGER1_TOKEN);
+    PayingApi api = new PayingApi(manager1Client);
+
+    Fee deletedFee = api.deleteStudentFeeById(FEE1_ID, STUDENT1_ID);
+    assertEquals(fee1(), deletedFee);
+
+    List<Fee> fees = api.getStudentFees(STUDENT1_ID, 1, 5, null);
+    assertTrue(!fees.contains(deletedFee));
```
Comment (Mahefaa) [inline src/test/java/school/hei/haapi/integration/FeeIT.java]: use assertFalse

File: src/main/java/school/hei/haapi/service/GroupService.java [hunk 14]
```diff
@@ -28,6 +28,12 @@ public class GroupService {
...
 
+  public Group deleteById(String groupId) {
+    Group group = getById(groupId);
+    repository.deleteById(groupId);
+    return group;
+  }
+
   public Group getById(String groupId) {
     return repository.getById(groupId);
   }
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/GroupService.java]: use findBy and orElseThrow

File: src/test/java/school/hei/haapi/integration/StudentIT.java [hunk 15]
```diff
@@ -278,6 +279,27 @@ void student_update_own_ok() throws ApiException {
     assertTrue(actualStudents.contains(actual));
   }
 
+  @Test
+  void student_delete_other_ko() throws ApiException {
+    ApiClient student1Client = anApiClient(STUDENT1_TOKEN);
+    UsersApi api = new UsersApi(student1Client);
+
+    assertThrowsForbiddenException(() -> api.deleteStudentById(STUDENT3_ID));
+  }
```
Comment (Mahefaa) [inline src/test/java/school/hei/haapi/integration/StudentIT.java]: student _delete_ko whether its their id or not

**Review by Mahefaa — CHANGES_REQUESTED** [review]
assertFalse exists x)

File: src/main/java/school/hei/haapi/service/GroupService.java (hunk 14 above)
Comment (haritianaadriano) [reply src/main/java/school/hei/haapi/service/GroupService.java]: but it's not redundant ?

File: doc/api.yml [hunk 16]
```diff
@@ -1102,6 +1160,34 @@ paths:
           $ref: '#/components/responses/429'
         '500':
           $ref: '#/components/responses/500'
+    delete:
+      tags:
+        - Users
+      summary: Delete teacher by identifier.
+      operationId: deleteTeacherById
```
Comment (Mahefaa) [inline doc/api.yml:1167]: description: soft delete

File: src/main/java/school/hei/haapi/model/Course.java [hunk 17]
```diff
@@ -34,6 +44,8 @@ public class Course implements Serializable {
 
   private Integer totalHours;
 
+  private boolean isDeleted = false;
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/model/Course.java]: why = false !

**Review by Mahefaa — CHANGES_REQUESTED** [review]
what will happen to a student's resources if the student gets deleted ?

File: src/main/java/school/hei/haapi/model/Course.java (hunk 17 above)
Comment (haritianaadriano) [reply src/main/java/school/hei/haapi/model/Course.java]: you said that we need to set it to false by default

File: doc/api.yml [hunk 18]
```diff
@@ -404,6 +433,34 @@ paths:
           $ref: '#/components/responses/429'
         '500':
           $ref: '#/components/responses/500'
+    delete:
+      tags:
+        - Teaching
+      summary: Delete course by identifier.
+      operationId: deleteCourseById
```
Comment (Mahefaa) [inline doc/api.yml:440]: description for soft delete

File: src/main/java/school/hei/haapi/model/GroupFlow.java [hunk 19]
```diff
@@ -36,18 +42,22 @@
...
 public class GroupFlow implements Serializable {
   // todo: to review all class
   @Id
   @GeneratedValue(strategy = IDENTITY)
   private String id;
 
-  @ManyToOne(fetch = LAZY)
+  @ManyToOne(fetch = LAZY, cascade = REMOVE)
   @JoinColumn(name = "student_id")
+  @OnDelete(action = CASCADE)
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/model/GroupFlow.java]: Deleting a groupFlow will delete the related student ??

**Review by Mahefaa — CHANGES_REQUESTED** [review]
it definitely doesn't tell me what happens to child entities on parent entity removal.

Also, have you tested whether the is_deleted variable really changes ?

**Pull request closed without merging**